src/provider/koios.ts

```diff
@@ -0,0 +1,221 @@
+            minFeeB: parseInt(result[0].min_fee_b),
+            maxTxSize: parseInt(result[0].max_tx_size),
+            maxValSize: parseInt(result[0].max_val_size),
+            keyDeposit: BigInt(result[0].key_deposit),
+            poolDeposit: BigInt(result[0].pool_deposit),
+            priceMem: parseFloat(result[0].price_mem),
+            priceStep: parseFloat(result[0].price_step),
+            maxTxExMem: BigInt(result[0].max_tx_ex_mem),
+            maxTxExSteps: BigInt(result[0].max_tx_ex_steps),
+            coinsPerUtxoByte: BigInt(result[0].coins_per_utxo_size),
+            collateralPercentage: parseInt(result[0].collateral_percent),
+            maxCollateralInputs: parseInt(result[0].max_collateral_inputs),
+            costModels: JSON.parse(result[0].cost_models),
+        };
+    }
+
+    async getUtxos(addressOrCredential: Address | Credential): Promise<UTxO[]> {
+        if (typeof addressOrCredential === "string") {
+            const result = await this.backendService.getAddressService().getAddressInformation([addressOrCredential]);
+            if (Array.isArray(result) && result.length > 0 && result[0].utxo_set && result[0].utxo_set.length > 0) {
+                return (await this.koiosUtxosToUtxos(result[0].utxo_set, result[0].address))
+            } else {
+                return []
+            }
+        } else {
+            throw Error('getUtxos by Credential Type is not supported in Koios yet.')
+        }
+    }
+
+    private async koiosUtxosToUtxos(result: any, address?: string): Promise<UTxO[]> {
+        return (await Promise.all(
+            result.map(async (r: any) => ({
+                txHash: r.tx_hash,
+                outputIndex: r.tx_index,
+                assets: (() => {
+                    const a: Assets = {};
+                    r.asset_list.forEach((am: any) => {
+                        a[am.policy_id + am.asset_name] = BigInt(am.quantity);
+                    });
+                    return a;
+                })(),
+                address: address ? address : r.payment_addr.bech32,
+                datumHash: !r.inline_datum ? r.datum_hash : undefined,
+                datum: r.inline_datum,
+                scriptRef: {
+                    type: r.reference_script ? r.reference_script.type : null,
+                    script: r.reference_script ? applyDoubleCborEncoding(r.reference_script.bytes) : null
+                },
+            })),
+        )) as UTxO[];
+    }
+
+    async getUtxosWithUnit(addressOrCredential: Address | Credential, unit: Unit): Promise<UTxO[]> {
+        if (typeof addressOrCredential === "string") {
+            const utxos = await this.getUtxos(addressOrCredential);
+            if (utxos && utxos.length > 0) {
+                return utxos.filter((utxo): utxo is UTxO => {
+                    const keys = Object.keys(utxo.assets)
+                    return keys.length > 0 && keys.includes(unit)
+                })
+            } else {
```

Comment:
This seems quite expensive. The point of the `getUtxosWithUnit` and `getUtxoByUnit` endpoint is that you can ask for a filtered UTxO set. These endpoints are usually used to query certain UTxOs from script addresses. Now imagine they have thousands of UTxOs. If you send all of them first and then filter locally in Lucid, it's gonna be slow.

Reply:
Since we don't have it in Koios that's the best it can get.
Isn't that what you meant in #102 `getUtxosWithUnit` explanation?

> If you provider doesn't have such an endpoint, just query all UTxOs and filter by the asset in the getUtxosWithUnit function in Lucid itself.

Reply:
Yes, I'm just saying it might be slow. It's a very useful endpoint when building dApps.

Reply:
let's leave it as it is for now and we will improve it later on as Koios roadmap includes major improvements in the future.
In General, Koios is designed to handle bulk queries for multiple addresses and only then drill down for the particular info you seek using `_info` endpoints.

Please feel free to review Koios [API ](https://api.koios.rest/) and let me know if you need some more capabilities.